src/ui/public/visualize/visualization.js

```diff
@@ -17,137 +17,68 @@
-          debounceTime(100),
-          switchMap(async ({ vis, visData, container }) => {
-            vis.size = [container.width(), container.height()];
-            const status = getUpdateStatus(vis.type.requiresUpdateStatus, $scope);
-            const renderPromise = visualization.render(visData, status);
-            $scope.$apply();
-            return renderPromise;
-          })
-        );
-
-        const requestError$ = render$.pipe(
-          filter(({ vis }) => vis.requestError)
-        );
-
-        const renderSubscription = Rx.merge(success$, requestError$)
-          .subscribe(() => {
-            dispatchRenderComplete($el[0]);
-          });
-
-        $scope.$on('$destroy', () => {
-          visualization.destroy();
-          renderSubscription.unsubscribe();
-        });
-
-        // Listen on changes to trigger a render if listenOnChange is not false
-        // i.e. true or has not been used (undefined in that case)
-        if ($scope.listenOnChange !== false) {
-          const onChangeListener = () => {
-            $scope.$emit('render');
-          };
-
-          $scope.$watchGroup(['visData', 'vis.params'], onChangeListener);
-
-          const resizeChecker = new ResizeChecker($el);
-          resizeChecker.on('resize', () => {
-            $scope.$emit('render');
-          });
-
-          $scope.uiState.on('change', onChangeListener);
-          $scope.$on('$destroy', () => {
-            resizeChecker.destroy();
-            $scope.uiState.off('change', onChangeListener);
-          });
-        }
-
-        function jQueryGetter(selector) {
-          return function () {
-            const $sel = $el.find(selector);
-            if ($sel.length) return $sel;
-          };
-        }
-      }
+import React, { Component } from 'react';
+import { VisualizationNoResults, VisualizationChart } from './components';
+
+const _showNoResultsMessage = (vis, visData) => {
+  const requiresSearch = _.get(vis, 'type.requiresSearch');
+  const isZeroHits = _.get(visData, 'hits.total') === 0;
+  const shouldShowMessage = !_.get(vis, 'params.handleNoResults');
+
+  return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+};
+
+export class Visualization extends Component {
+  constructor(props) {
+    super(props);
+    props.vis.initialized = true;
+
+    props.vis._setUiState(props.uiState);
+    if (props.listenOnChange) {
+      props.uiState.on('change', this._onChangeListener);
+    }
+
+    this.state = {
+      listenOnChange: props.listenOnChange,
+      showNoResultsMessage: _showNoResultsMessage(props.vis, props.visData)
     };
-  });
+  }
+
+  render() {
+    return (
+      <div className="visualization">
+        {this.state.showNoResultsMessage ? (<VisualizationNoResults />) :
+          (<VisualizationChart
+            vis={this.props.vis}
+            visData={this.props.visData}
+            listenOnChange={this.props.listenOnChange}
```

Comment:
From our [styleguide](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#use-object-destructuring), it's better to deconstruct objects early as possible:
`const { vis, visData, listenOnChange } = this.props`